File: src/main/java/core/procedures/strings/StringCopy.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMMutableString;

public final class StringCopy extends AFn {

  public StringCopy() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{CharSequence.class}).build());
  }

  @Override
  public String getName() {
    return "string-copy";
  }

  @Override
  public SCMMutableString apply1(Object arg) {
    return new SCMMutableString(arg.toString());
  }
}

File: src/main/java/core/procedures/strings/StringCopy.kt
package core.procedures.strings

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.MutableString

class StringCopy : AFn<CharSequence?, MutableString>(name = "string-copy", isPure = true, arity = Exactly(1),
                                                     mandatoryArgsTypes = arrayOf(CharSequence::class.java)) {

    override operator fun invoke(arg: CharSequence?) = MutableString(arg!!.toString())
}
